# Fannie Mae - Defaulter Prediction - ANN

In [1]:
import keras
from keras.models import Sequential
from keras.layers import Dense
import numpy as np
import pandas as pd

# fix random seed for reproducibility
np.random.seed(7)


Using TensorFlow backend.


In [2]:
# Original imbalanced dataset after Cleaning and feature generation
dataset = pd.read_csv("C:\\DATASCIENCE\\11-DONO-PROJECT\\2-Jigsaw_Dono\\DATASTORE\\mbs_scaled_final1234.csv", sep='\t', index_col=0)
# split into input (X) and output (Y) variables
X = dataset.drop("DEFAULT STATUS", axis=1)
Y = dataset["DEFAULT STATUS"]

C:\Users\Prabha\Anaconda3\envs\Deeplearning\lib\site-packages\numpy\lib\arraysetops.py:569: FutureWarning: elementwise comparison failed; returning scalar instead, but in the future will perform elementwise comparison
  mask |= (ar1 == a)


In [3]:
X.shape

(1807092, 98)

In [4]:
Y.shape

(1807092,)

In [5]:
# create model
model = Sequential()
model.add(Dense(12, input_dim=98, activation='relu'))
model.add(Dense(8, activation='relu'))
model.add(Dense(1, activation='sigmoid'))

Instructions for updating:
Colocations handled automatically by placer.


In [6]:
# Compile model
model.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])

In [7]:
model.fit(X, Y, epochs=10, batch_size=10000 )

Instructions for updating:
Use tf.cast instead.
Epoch 1/10
1807092/1807092 [==============================] - 16s 9us/step - loss: 0.1397 - acc: 0.9949
Epoch 2/10
1807092/1807092 [==============================] - 15s 8us/step - loss: 0.0284 - acc: 0.9953: 2s - lo - ETA: 0s - loss: 0.0284 - acc: 0
Epoch 3/10
1807092/1807092 [==============================] - 15s 8us/step - loss: 0.0264 - acc: 0.9953
Epoch 4/10
1807092/1807092 [==============================] - 15s 9us/step - loss: 0.0257 - acc: 0.9953: 0s - loss: 0.0257 - acc: 0
Epoch 5/10
1807092/1807092 [==============================] - 16s 9us/step - loss: 0.0253 - acc: 0.9953
Epoch 6/10
1807092/1807092 [==============================] - 16s 9us/step - loss: 0.0251 - acc: 0.9953
Epoch 7/10
1807092/1807092 [==============================] - 15s 8us/step - loss: 0.0250 - acc: 0.9953
Epoch 8/10
1807092/1807092 [==============================] - 15s 8us/step - loss: 0.0249 - acc: 0.9953
Epoch 9/10
1807092/1807092 [=====================

In [8]:
# Test dataset
Xu= pd.read_csv("C:\\DATASCIENCE\\11-DONO-PROJECT\\2-Jigsaw_Dono\\DATASTORE\\X_testoversample.csv", sep='\t', index_col=0)
Yu=pd.read_csv("C:\\DATASCIENCE\\11-DONO-PROJECT\\2-Jigsaw_Dono\\DATASTORE\\y_testoversample.csv", sep='\t', index_col=0)

In [9]:
# evaluate the model
scores = model.evaluate(Xu, Yu)
print("\n%s: %.2f%%" % (model.metrics_names[1], scores[1]*100))

361419/361419 [==============================] - 10s 28us/step

acc: 99.52%


### ANN with oversampled dataset

In [10]:
# Over sampled dataset
Xo= pd.read_csv("C:\\DATASCIENCE\\11-DONO-PROJECT\\2-Jigsaw_Dono\\DATASTORE\\X_trainoversample.csv", sep='\t', index_col=0)
Yo=pd.read_csv("C:\\DATASCIENCE\\11-DONO-PROJECT\\2-Jigsaw_Dono\\DATASTORE\\y_trainoversample.csv", sep='\t', index_col=0)

In [11]:
Xo.shape

(2877846, 98)

In [12]:
Yo.shape

(2877846, 1)

In [13]:
# create model
model = Sequential()
model.add(Dense(12, input_dim=98, activation='relu'))
model.add(Dense(8, activation='relu'))
model.add(Dense(4, activation='relu'))
model.add(Dense(1, activation='sigmoid'))

In [14]:
# Compile model
model.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])

In [15]:
model.fit(Xo, Yo, epochs=20, batch_size=10000 )

Epoch 1/20
2877846/2877846 [==============================] - 27s 9us/step - loss: 0.5098 - acc: 0.7665
Epoch 2/20
2877846/2877846 [==============================] - 27s 9us/step - loss: 0.3813 - acc: 0.8326
Epoch 3/20
2877846/2877846 [==============================] - 27s 9us/step - loss: 0.3560 - acc: 0.8466
Epoch 4/20
2877846/2877846 [==============================] - 27s 9us/step - loss: 0.3273 - acc: 0.8614: 12s - loss: 0. - ETA: 11s - loss: 0.3328 - acc: 0. - ETA: 11 - ETA: - ETA:
Epoch 5/20
2877846/2877846 [==============================] - 27s 9us/step - loss: 0.3068 - acc: 0.8716
Epoch 6/20
2877846/2877846 [==============================] - 27s 9us/step - loss: 0.2934 - acc: 0.8773
Epoch 7/20
2877846/2877846 [==============================] - 27s 9us/step - loss: 0.2834 - acc: 0.8820
Epoch 8/20
2877846/2877846 [==============================] - 27s 9us/step - loss: 0.2759 - acc: 0.8855: 8s - loss: 0.2769 - acc - ETA: 0s - loss: 0.2760 - acc: 0.885
Epoch 9/20
2877846/2877846 [=

In [16]:
# evaluate the model on undersampled dataset
scores = model.evaluate(Xu, Yu)
print("\n%s: %.2f%%" % (model.metrics_names[1], scores[1]*100))

361419/361419 [==============================] - 11s 30us/step

acc: 88.77%


### Artificial Neural Networks
The actual data and oversampled data were used to make the predictions. 

As the actual dataset had a significantly fewer Defaulters (‘1’  : 0.47%) as opposed to ‘0”. The prediction accuracy for this dataset was close to 99.5%. The accuracy was driven by the fact that almost all the predicted values were 0. The same is the case with test data also as it also has very low % of defaulters.

In order to overcome this bias in the prediction, we tried a modeling approach using Oversampled data. Here though the accuracy on train data reduced to 89% compared to previous model, the accuracy with test data increased to 88%. While this model has lesser accuracy than the previous model, this is a more realistic predictions. 